src/Stratis.Bitcoin.Features.Api/MvcBuilderExtensions.cs

```diff
@@ -1,37 +1,32 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
-using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.DependencyInjection;
-using Stratis.Bitcoin.Controllers;
+using Stratis.Bitcoin.Builder.Feature;
 
 namespace Stratis.Bitcoin.Features.Api
 {
     public static class MvcBuilderExtensions
     {
         /// <summary>
-        /// Finds all the types that are <see cref="Controller"/> or <see cref="FeatureController"/>and add them to the Api as services.
+        /// Identifies the assemblies to include for controller discovery.
         /// </summary>
         /// <param name="builder">The builder</param>
-        /// <param name="services">The services to look into</param>
+        /// <param name="features">The selected features to include the assemblies of.</param>
         /// <returns>The Mvc builder</returns>
-        public static IMvcBuilder AddControllers(this IMvcBuilder builder, IServiceCollection services)
+        public static IMvcBuilder AddControllers(this IMvcBuilder builder, IEnumerable<IFullNodeFeature> features)
         {
-            // Adds Controllers with API endpoints
-            System.Collections.Generic.IEnumerable<ServiceDescriptor> controllerTypes = services.Where(s => s.ServiceType.GetTypeInfo().BaseType == typeof(Controller));
-            foreach (ServiceDescriptor controllerType in controllerTypes)
-            {
-                builder.AddApplicationPart(controllerType.ServiceType.GetTypeInfo().Assembly);
-            }
+            // The required assemblies including those of the selected features.
+            IEnumerable<Assembly> assemblies = features
+                .Select(f => f.GetType().Assembly)
+                .Append(typeof(Controllers.NodeController).Assembly)
+                .Distinct();
 
-            // Adds FeatureControllers with API endpoints.
-            System.Collections.Generic.IEnumerable<ServiceDescriptor> featureControllerTypes = services.Where(s => s.ServiceType.GetTypeInfo().BaseType == typeof(FeatureController));
-            foreach (ServiceDescriptor featureControllerType in featureControllerTypes)
+            foreach (Assembly assembly in assemblies)
             {
-                builder.AddApplicationPart(featureControllerType.ServiceType.GetTypeInfo().Assembly);
+                builder.AddApplicationPart(assembly);
```

Comment:
I think you missed the other assemblies here? Because up top you only look for the `NodeController`'s assembly? I think in this way you are not adding all the other feature's controllers? Sorry if I'm being blind here.

Reply:
The `Select` selects the feature assemblies. The reason for having `NodeController` in addition to this is that it would be missed as its not a feature.

Reply:
Ah my bad!